calblueprint/girls-write-now
# [rec] Implement recommendation algorithm

[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

- Implemented a recommendation algorithm that displays the user's recommended stories on the home page
- Recommended stories are based on the user's recently viewed stories, finding the 5 most similar stories to the recently viewed stories based on the story's content
- Recommendations only refresh once the home page is refreshed
- Added functionality to add stories on the home page to the recently viewed when clicking on them to keep the recommendation algorithm updated 

## Relevant Links

### Notion Sprint Task

[//]: # 'Add the relevant Notion link(s) here'

https://www.notion.so/calblueprint/rec-Implement-recommendation-algorithm-973d28ff52ad4357aa94e851032bebf8

### Online sources

[//]: # 'Optional - copy links to any tutorial or documentation that was useful to you when working on this PR'

Supabase OpenAI Vector extension documentation: 
- https://supabase.com/docs/guides/ai/vector-columns#querying-a-vector--embedding
- https://supabase.com/docs/guides/database/extensions/pgvector

Tutorial the algorithm is mostly based on: https://www.notjust.dev/blog/ai-movies-supabase-vector

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

## How to review

[//]: # 'The order in which to review files and what to expect when testing locally'

- Open the home page and check if there are recommended stories, if not, open some random stories and refresh the app to see if recommended stories are displayed
- Add more stories to your recently viewed by searching for them and clicking on stories on the home page
- Refresh again and check that different stories are recommended 
- Check the recommended stories to see if they are actually (somewhat) similar to your recently viewed stories

## Next steps

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

- Added functions to the Node which still needs to be pushed

## Tests Performed, Edge Cases

[//]: # 'Hopefully we will add a testing suite/CI soon, but until then note down the steps you took to test locally'

- Checked that stories are recommended if there are stories in the recently viewed
- Checked that no stories are recommended if there are no recently viewed stories
- Checked that the recently viewed stories in the search screen still works with these new changes 

### Screenshots

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"


https://github.com/calblueprint/girls-write-now/assets/156966341/f6450761-53aa-4113-a338-5efe4d32ec72


CC: @adityapawar1

[//]: # 'This tags in Adi as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
